source/drivers/Sensor.cpp

```diff
@@ -0,0 +1,247 @@
+ * Creates a generic Sensor.
+ *
+ * @param id The ID of this compoenent e.g. DEVICE_ID_THERMOMETER
+ */
+Sensor::Sensor(uint16_t id)
+{
+    this->id = id;
+    this->sensitivity = 868;
+
+    // Configure for a 2 Hz update frequency by default.
+    if(EventModel::defaultEventBus)
+        EventModel::defaultEventBus->listen(id, SENSOR_UPDATE_NEEDED, this, &Sensor::onSampleEvent, MESSAGE_BUS_LISTENER_IMMEDIATE);
+
+    setPeriod(500);
+}
+
+/*
+ * Event Handler for periodic sample timer
+ */
+void Sensor::onSampleEvent(Event)
+{
+    updateSample();
+}
+
+/*
+ * Determines the instantaneous value of the sensor, in SI units, and returns it.
+ *
+ * @return The current value of the sensor.
+ */
+int Sensor::getValue()
+{
+    return (int)sensorValue;
+}
+
+/**
+ * Updates the internal reading of the sensor. Typically called periodically.
+ */
+void Sensor::updateSample()
+{
+    uint32_t value = readValue();
+
+    // If this is the first reading performed, take it a a baseline. Otherwise, perform a decay average to smooth out the data.
+    if (!(status & SENSOR_INITIALISED))
+    {
+        sensorValue = (uint16_t)value;
+        status |=  SENSOR_INITIALISED;
+        // set Period to default value
+        setPeriod(500);
+    }
+    else
+    {
+        sensorValue = ((sensorValue * (1023 - sensitivity)) + (value * sensitivity)) >> 10;
+    }
+
+    checkThresholding();
+}
+
+/**
+ * Determine if any thresholding events need to be generated, and if so, raise them.
+ */
+void Sensor::checkThresholding()
+{
+    if ((status & SENSOR_HIGH_THRESHOLD_ENABLED) && (!(status & SENSOR_HIGH_THRESHOLD_PASSED)) && (sensorValue >= highThreshold))
+    {
+        Event(id, SENSOR_THRESHOLD_HIGH);
+        status |=  SENSOR_HIGH_THRESHOLD_PASSED;
+        status &= ~SENSOR_LOW_THRESHOLD_PASSED;
+    }
+
+    if ((status & SENSOR_LOW_THRESHOLD_ENABLED) && (!(status & SENSOR_LOW_THRESHOLD_PASSED)) && (sensorValue <= lowThreshold))
+
+    {
+        Event(id, SENSOR_THRESHOLD_LOW);
+        status |=  SENSOR_LOW_THRESHOLD_PASSED;
+        status &= ~SENSOR_HIGH_THRESHOLD_PASSED;
+    }
+}
+
+/**
+ * Set sensitivity value for the data. A decay average is taken of sampled data to smooth it into more accurate information.
+ *
+ * @param value A value between 0..1023 that detemrines the level of smoothing. Set to 1023 to disable smoothing. Default value is 868
+ *
+ * @return DEVICE_OK on success, DEVICE_INVALID_PARAMETER if the request fails.
+ */
+int Sensor::setSensitivity(uint16_t value)
+{
+    this->sensitivity = max(0, min(1023, value));
+
+    return DEVICE_OK;
+}
+
+/**
+ * Set the automatic sample period of the sensor to the specified value (in ms).
+ *
+ * @param period the requested time between samples, in milliseconds.
+ *
+ * @return DEVICE_OK on success, DEVICE_INVALID_PARAMETER if the request fails.
+ */
+int Sensor::setPeriod(int period)
+{
+    samplePeriod = period;
+    system_timer_event_every_us(1000 * period, id, SENSOR_UPDATE_NEEDED);
```

Comment:
Could just use `system_timer_event_every` (for milliseconds) rather than the `_us` variant

Reply:
Corrected !